File: design_pattern/observer.ts
// ============================================================
// Observer Pattern — 状态变化时自动通知所有订阅者
// 对比 Rust: 08_observer.rs
// 运行: npx ts-node 08_observer.ts
// ============================================================

type Handler<T> = (event: T) => void;

class EventEmitter<T> {
  private handlers = new Map<number, Handler<T>>();
  private nextId = 0;

  on(handler: Handler<T>): number {
    const id = this.nextId++;
    this.handlers.set(id, handler);
    return id;
  }

  off(id: number) {
    this.handlers.delete(id);
    console.log(`[Emitter] 取消订阅 id=${id}`);
  }

  emit(event: T) {
    this.handlers.forEach(h => h(event));
  }
}

interface UserEvent {
  userId: number;
  action: string;
}

// --- main ---
console.log("=== Observer Pattern ===\n");

const emitter = new EventEmitter<UserEvent>();

const logId = emitter.on(e => {
  console.log(`[Log]   user=${e.userId} action=${e.action}`);
});

const _statId = emitter.on(e => {
  console.log(`[Stats] 记录事件: ${e.action}`);
});

const emailId = emitter.on(e => {
  if (e.action === "register") {
    console.log(`[Email] 发送欢迎邮件给 user=${e.userId}`);
  }
});

console.log("--- 触发 register ---");
emitter.emit({ userId: 1, action: "register" });

console.log("\n--- 触发 login ---");
emitter.emit({ userId: 1, action: "login" });

console.log("\n--- 取消 email + log 订阅 ---");
emitter.off(emailId);
emitter.off(logId);

console.log("\n--- 再次触发（只剩 stats）---");
emitter.emit({ userId: 2, action: "login" });

File: design_pattern/observer.rs
// ============================================================
// Observer Pattern — 状态变化时自动通知所有订阅者
// 对比 TS: 08_observer.ts
// 运行: cargo run --bin observer
// ============================================================

use std::collections::HashMap;

type HandlerId = u64;

struct EventEmitter<T> {
    next_id: HandlerId,
    handlers: HashMap<HandlerId, Box<dyn Fn(&T)>>,
}

impl<T> EventEmitter<T> {
    fn new() -> Self {
        Self { next_id: 0, handlers: HashMap::new() }
    }

    fn on(&mut self, handler: impl Fn(&T) + 'static) -> HandlerId {
        let id = self.next_id;
        self.next_id += 1;
        self.handlers.insert(id, Box::new(handler));
        id
    }

    fn off(&mut self, id: HandlerId) {
        self.handlers.remove(&id);
        println!("[Emitter] 取消订阅 id={}", id);
    }

    fn emit(&self, event: &T) {
        for handler in self.handlers.values() {
            handler(event);
        }
    }
}

#[derive(Debug)]
struct UserEvent {
    user_id: u64,
    action: String,
}

fn main() {
    println!("=== Observer Pattern ===\n");

    let mut emitter = EventEmitter::<UserEvent>::new();

    let log_id = emitter.on(|e| {
        println!("[Log]   user={} action={}", e.user_id, e.action);
    });

    let _stat_id = emitter.on(|e| {
        println!("[Stats] 记录事件: {}", e.action);
    });

    let email_id = emitter.on(|e| {
        if e.action == "register" {
            println!("[Email] 发送欢迎邮件给 user={}", e.user_id);
        }
    });

    println!("--- 触发 register ---");
    emitter.emit(&UserEvent { user_id: 1, action: "register".into() });

    println!("\n--- 触发 login ---");
    emitter.emit(&UserEvent { user_id: 1, action: "login".into() });

    println!("\n--- 取消 email + log 订阅 ---");
    emitter.off(email_id);
    emitter.off(log_id);

    println!("\n--- 再次触发（只剩 stats）---");
    emitter.emit(&UserEvent { user_id: 2, action: "login".into() });
}
